File: WinForms/VB/HtmlMainDemo/Modules/AccordionControl/HamburgerMenuModule.vb
Imports DevExpress.DXperience.Demos
Imports DevExpress.LookAndFeel
Imports DevExpress.XtraBars.Navigation
Imports System.ComponentModel
Imports System.Drawing

Namespace DevExpress.HTML.Demos

    Public Partial Class HamburgerMenuModule
        Inherits TutorialControlBase

        Public Sub New()
            InitializeComponent()
            accordionControlTemplateCodeViewer1.Fill(accordionControl1, Nothing)
            UpdateBackground()
            InitCategories()
        End Sub

        Protected Overrides Sub OnLookAndFeelChanged()
            MyBase.OnLookAndFeelChanged()
            UpdateBackground()
        End Sub

        Private Sub UpdateBackground()
            Dim bg As Color = LookAndFeelHelper.GetSystemColor(LookAndFeel.ActiveLookAndFeel, SystemColors.Control)
            accordionControl1.Appearance.AccordionControl.BackColor = bg
            BackColor = bg
        End Sub

        Private Sub InitCategories()
            acEmployees.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Salaried"), New HamburgerMenuCategory("Commission"), New HamburgerMenuCategory("Contract")}, New HamburgerMenuCategory() {New HamburgerMenuCategory("Terminated"), New HamburgerMenuCategory("On Leave")})
            acCustomers.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Stores > 10"), New HamburgerMenuCategory("Revenue > 100 Billion")}, New HamburgerMenuCategory() {New HamburgerMenuCategory("Employees > 10000")})
            acVendors.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Active"), New HamburgerMenuCategory("Suspended")}, Nothing)
            acMonitors.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Available")}, New HamburgerMenuCategory() {New HamburgerMenuCategory("Discounted")})
            acProjectors.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Available")}, New HamburgerMenuCategory() {New HamburgerMenuCategory("Discounted")})
            acTelevisions.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Available")}, New HamburgerMenuCategory() {New HamburgerMenuCategory("Discounted")})
            acVideoPlayers.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Available")}, New HamburgerMenuCategory() {New HamburgerMenuCategory("Discounted")})
            acMonitors2.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Today"), New HamburgerMenuCategory("Yesterday"), New HamburgerMenuCategory("This Month")}, New HamburgerMenuCategory() {New HamburgerMenuCategory("This Year"), New HamburgerMenuCategory("Unpaid Orders")})
            acProjectors2.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Today"), New HamburgerMenuCategory("Yesterday"), New HamburgerMenuCategory("This Month")}, New HamburgerMenuCategory() {New HamburgerMenuCategory("This Year"), New HamburgerMenuCategory("Unpaid Orders")})
            acTelevisions2.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Today"), New HamburgerMenuCategory("Yesterday"), New HamburgerMenuCategory("This Month")}, New HamburgerMenuCategory() {New HamburgerMenuCategory("This Year"), New HamburgerMenuCategory("Unpaid Orders")})
            acVideoPlayers2.Tag = New HamburgerElementDataSource(New HamburgerMenuCategory() {New HamburgerMenuCategory("Today"), New HamburgerMenuCategory("Yesterday"), New HamburgerMenuCategory("This Month")}, New HamburgerMenuCategory() {New HamburgerMenuCategory("This Year"), New HamburgerMenuCategory("Unpaid Orders")})
        End Sub

        Private Sub accordionControl1_QueryHtmlElementData(ByVal sender As Object, ByVal e As QueryAccordionHtmlElementDataEventArgs)
            If Equals(e.FieldName, "IsControlExpanded") Then
                e.Value = CType(sender, AccordionControl).OptionsMinimizing.State = AccordionControlState.Normal
                Return
            End If

            If e.Element Is Nothing Then Return
            If Equals(e.FieldName, "Categories1") Then
                Dim data = TryCast(e.Element.Tag, HamburgerElementDataSource)
                e.Value = data?.Categories1
                Return
            End If

            If Equals(e.FieldName, "Categories2") Then
                Dim data = TryCast(e.Element.Tag, HamburgerElementDataSource)
                e.Value = data?.Categories2
                Return
            End If

            If Equals(e.FieldName, "IsFirst") Then
                Dim elements = e.Element.OwnerElement?.Elements
                e.Value = elements IsNot Nothing AndAlso elements.IndexOf(e.Element) = 0
            End If

            If Equals(e.FieldName, "IsLast") Then
                Dim elements = e.Element.OwnerElement?.Elements
                e.Value = elements IsNot Nothing AndAlso elements.IndexOf(e.Element) = elements.Count - 1
            End If
        End Sub

        Private Sub accordionControl1_HtmlElementMouseClick(ByVal sender As Object, ByVal e As AccordionHtmlElementEventArgs)
            If e.HasId("searchbutton") Then
                Dim accordion = CType(sender, AccordionControl)
                If accordion.ShowFilterControl = ShowFilterControl.Never Then
                    accordion.ShowFilterControl = ShowFilterControl.Always
                Else
                    accordion.ShowFilterControl = ShowFilterControl.Never
                End If
            End If
        End Sub
    End Class

    Friend Class HamburgerElementDataSource

        Public Sub New(ByVal categories1 As HamburgerMenuCategory(), ByVal categories2 As HamburgerMenuCategory())
            Me.Categories1 = categories1
            Me.Categories2 = categories2
        End Sub

        Public Property Categories1 As HamburgerMenuCategory()

        Public Property Categories2 As HamburgerMenuCategory()
    End Class

    Friend Class HamburgerMenuCategory

        Public Sub New(ByVal name As String)
            Me.Name = name
        End Sub

        Public Property Name As String
    End Class
End Namespace

File: WinForms/CS/HtmlMainDemo/Modules/AccordionControl/HamburgerMenuModule.cs
using DevExpress.DXperience.Demos;
using DevExpress.LookAndFeel;
using DevExpress.XtraBars.Navigation;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace DevExpress.HTML.Demos {
    public partial class HamburgerMenuModule : TutorialControlBase {
        public HamburgerMenuModule() {
            InitializeComponent();
            this.accordionControlTemplateCodeViewer1.Fill(this.accordionControl1, null);
            UpdateBackground();
            InitCategories();
        }
        protected override void OnLookAndFeelChanged() {
            base.OnLookAndFeelChanged();
            UpdateBackground();
        }
        void UpdateBackground() {
            Color bg = LookAndFeelHelper.GetSystemColor(LookAndFeel.ActiveLookAndFeel, SystemColors.Control);
            this.accordionControl1.Appearance.AccordionControl.BackColor = bg;
            BackColor = bg;
        }
        void InitCategories() {
            this.acEmployees.Tag = new HamburgerElementDataSource(
                new HamburgerMenuCategory[] { new HamburgerMenuCategory("Salaried"), new HamburgerMenuCategory("Commission"), new HamburgerMenuCategory("Contract") },
                new HamburgerMenuCategory[] { new HamburgerMenuCategory("Terminated"), new HamburgerMenuCategory("On Leave") }
                );
            this.acCustomers.Tag = new HamburgerElementDataSource(
                new HamburgerMenuCategory[] { new HamburgerMenuCategory("Stores > 10"), new HamburgerMenuCategory("Revenue > 100 Billion") },
                new HamburgerMenuCategory[] { new HamburgerMenuCategory("Employees > 10000") }
                );
            this.acVendors.Tag = new HamburgerElementDataSource(
                new HamburgerMenuCategory[] { new HamburgerMenuCategory("Active"), new HamburgerMenuCategory("Suspended") },
                null
                );
            this.acMonitors.Tag = new HamburgerElementDataSource(
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("Available") },
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("Discounted") }
               );
            this.acProjectors.Tag = new HamburgerElementDataSource(
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("Available") },
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("Discounted") }
               );
            this.acTelevisions.Tag = new HamburgerElementDataSource(
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("Available") },
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("Discounted") }
               );
            this.acVideoPlayers.Tag = new HamburgerElementDataSource(
                new HamburgerMenuCategory[] { new HamburgerMenuCategory("Available") },
                new HamburgerMenuCategory[] { new HamburgerMenuCategory("Discounted") }
                );
            this.acMonitors2.Tag = new HamburgerElementDataSource(
                new HamburgerMenuCategory[] { new HamburgerMenuCategory("Today"), new HamburgerMenuCategory("Yesterday"), new HamburgerMenuCategory("This Month") },
                new HamburgerMenuCategory[] { new HamburgerMenuCategory("This Year"), new HamburgerMenuCategory("Unpaid Orders") }
                );
            this.acProjectors2.Tag = new HamburgerElementDataSource(
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("Today"), new HamburgerMenuCategory("Yesterday"), new HamburgerMenuCategory("This Month") },
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("This Year"), new HamburgerMenuCategory("Unpaid Orders") }
               );
            this.acTelevisions2.Tag = new HamburgerElementDataSource(
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("Today"), new HamburgerMenuCategory("Yesterday"), new HamburgerMenuCategory("This Month") },
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("This Year"), new HamburgerMenuCategory("Unpaid Orders") }
               );
            this.acVideoPlayers2.Tag = new HamburgerElementDataSource(
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("Today"), new HamburgerMenuCategory("Yesterday"), new HamburgerMenuCategory("This Month") },
               new HamburgerMenuCategory[] { new HamburgerMenuCategory("This Year"), new HamburgerMenuCategory("Unpaid Orders") }
               );
        }

        private void accordionControl1_QueryHtmlElementData(object sender, XtraBars.Navigation.QueryAccordionHtmlElementDataEventArgs e) {
            if(e.FieldName == "IsControlExpanded") {
                e.Value = ((AccordionControl)sender).OptionsMinimizing.State == AccordionControlState.Normal;
                return;
            }
            if(e.Element == null)
                return;
            if(e.FieldName == "Categories1") {
                var data = e.Element.Tag as HamburgerElementDataSource;
                e.Value = data?.Categories1;
                return;
            }
            if(e.FieldName == "Categories2") {
                var data = e.Element.Tag as HamburgerElementDataSource;
                e.Value = data?.Categories2;
                return;
            }
            if(e.FieldName == "IsFirst") {
                var elements = e.Element.OwnerElement?.Elements;
                e.Value = elements != null && elements.IndexOf(e.Element) == 0;
            }
            if(e.FieldName == "IsLast") {
                var elements = e.Element.OwnerElement?.Elements;
                e.Value = elements != null && elements.IndexOf(e.Element) == elements.Count - 1;
            }
        }

        private void accordionControl1_HtmlElementMouseClick(object sender, AccordionHtmlElementEventArgs e) {
            if(e.HasId("searchbutton")) {
                var accordion = (AccordionControl)sender;
                if(accordion.ShowFilterControl == ShowFilterControl.Never)
                    accordion.ShowFilterControl = ShowFilterControl.Always;
                else
                    accordion.ShowFilterControl = ShowFilterControl.Never;
            }
        }
    }
    class HamburgerElementDataSource {
        public HamburgerElementDataSource(HamburgerMenuCategory[] categories1, HamburgerMenuCategory[] categories2) {
            Categories1 = categories1;
            Categories2 = categories2;
        }
        public HamburgerMenuCategory[] Categories1 { get; set; }
        public HamburgerMenuCategory[] Categories2 { get; set; }
    }
    class HamburgerMenuCategory {
        public HamburgerMenuCategory(string name) {
            Name = name;
        }
        public string Name { get; set; }
    }
}
